boot/makebootable.go

```diff
@@ -280,21 +280,42 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
 	blVars := map[string]string{
 		"kernel_status": "",
 	}
-	if err := bl.SetBootVars(blVars); err != nil {
-		return fmt.Errorf("cannot set run system environment: %v", err)
-	}
 
 	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
-	if !ok {
-		return fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", bl.Name())
-	}
+	if ok {
+		// we can use ebl, so just set the EnableKernel after setting
+		// kernel_status
+		if err := bl.SetBootVars(blVars); err != nil {
+			return fmt.Errorf("cannot set run system environment: %v", err)
+		}
 
-	if err := ebl.EnableKernel(bootWith.Kernel); err != nil {
-		return err
+		if err := ebl.EnableKernel(bootWith.Kernel); err != nil {
+			return err
+		}
+	} else {
+		// TODO:UC20: should we make this more explicit with a new
+		//            bootloader interface that is checked for first before
+		//            ExtractedRunKernelImageBootloader the same way we do with
+		//            ExtractedRecoveryKernelImageBootloader?
+
+		// we don't have an ExtractedRunKernelImageBootloader, so just set the
+		// legacy variables for snap_kernel and move on
+		blVars["snap_kernel"] = bootWith.Kernel.Filename()
+
+		if err := bl.SetBootVars(blVars); err != nil {
+			return fmt.Errorf("cannot set run system environment: %v", err)
+		}
 	}
 
 	// LAST step: update recovery grub's grubenv to indicate that
 	// we transition to run mode now
+
+	// TODO:UC20: this "just works" on accident with bootloaders like uboot and
+	//            lk, but should we make this more explicit? currently they
+	//            don't see or care about the Recovery option, so we in effect
+	//            end up writing the bootenv to the normal bootenv on
+	//            ubuntu-seed, which is what we wanted for these bootloaders
+	//            anyways?
```

Comment:
Recovery mostly is about the layout seen in the partition, this means we have /boot/uboot in the seed partition as well? is that what we want

Reply:
I think we can drop the TODO. The way I understand it, we ask to find a bootloader on the seed partition, letting it know that it will be acting on a recovery system. On amd64 it means that grub env is in different location (`/EFI/ubuntu`), while with uboot it's the same location `/boot/uboot`. Maybe it's an implementation detail leaking to the upper layers, but it's acceptable (at least for me :smile:).

Reply:
It's fine to drop this TODO now because uboot understands and uses Recovery here, so it's fine. I think it's well known about the fact that this may not be right for lk, but we also have a good friend who will make sure we are reminded to handle this for lk :smile: